test/blob.test.js

```diff
@@ -0,0 +1,251 @@
+
+  // Prepare data
+  const file = await randomFile(100)
+  
+  // Encode file as Unixfs and perform store/add
+  const blocksReadableStream = UnixFS.createFileEncoderStream(file)
+  /** @type {import('@web3-storage/upload-client/types').CARLink[]} */
+  const shards = []
+  /** @type {Uint8Array[]} */
+  const shardBytes = []
+  /** @type {Array<Map<import('@web3-storage/upload-client/types').SliceDigest, import('@web3-storage/upload-client/types').Position>>} */
+  const shardIndexes = []
+  /** @type {import('@web3-storage/upload-client/types').AnyLink | undefined} */
+  let root
+
+  /** @type {import('multiformats/hashes/digest').Digest<18, number> | undefined} */
+  let multihash
+  /** @type {{ put: any, accept: { task: any }} | undefined} */
+  let next
+  await blocksReadableStream
+    .pipeThrough(new ShardingStream())
+    .pipeThrough(
+      new TransformStream({
+        async transform(car, controller) {
+          const bytes = new Uint8Array(await car.arrayBuffer())
+
+          // Add blob using custom client to be able to access receipts
+          const res = await Blob.add(serviceProps.conf, bytes, { connection: serviceProps.connection })
+          t.truthy(res)
+          t.truthy(res.multihash)
+          multihash = res.multihash
+          next = res.next
+
+          const cid = Link.create(carCodec.code, res.multihash)
+          const { version, roots, size, slices } = car
+          controller.enqueue({ version, roots, size, cid, slices, bytes })
+        }
+      })
+    )
+    .pipeTo(
+      new WritableStream({
+        write(meta) {
+          root = root || meta.roots[0]
+          shards.push(meta.cid)
+          shardBytes.push(meta.bytes)
+
+          // add the CAR shard itself to the slices
+          meta.slices.set(meta.cid.multihash, [0, meta.size])
+          shardIndexes.push(meta.slices)
+        },
+      })
+    )
+
+  if (root === undefined) throw new Error('missing root CID')
+  if (multihash === undefined) throw new Error('missing multihash')
+  t.is(shards.length, 1)
+
+  console.log('root', root)
+
+  // Add the index with `index/add`
+  const index = ShardedDAGIndex.create(root)
+  for (const [i, shard] of shards.entries()) {
+    const slices = shardIndexes[i]
+    index.shards.set(shard.multihash, slices)
+  }
+  const indexBytes = await index.archive()
+  /* c8 ignore next 3 */
+  if (!indexBytes.ok) {
+    throw new Error('failed to archive DAG index', { cause: indexBytes.error })
+  }
+  // Store the index in the space
+  const resIndex = await Blob.add(serviceProps.conf, indexBytes.ok, { connection: serviceProps.connection })
+  const indexLink = Link.create(carCodec.code, resIndex.multihash)
+
+  // Register the index with the service
+  await Index.add(serviceProps.conf, indexLink, { connection: serviceProps.connection })
+  // Register an upload with the service
+  await Upload.add(serviceProps.conf, root, shards, { connection: serviceProps.connection })
```

Comment:
Why are we not using the `@web3-storage/w3up-client` for this?

Reply:
See comment above

```
// Add blob using custom client to be able to access receipts
```

I want to have the CIDs for the Receipts so that I can make sure they exist. The upload client is fully used on `integration.test.js`, this is a more internal testing. Also, this way, I have access to the CID of the index that I can also check is written

Reply:
I will improve docs though

Reply:
probably, I could also do a follow up where in the client we export a function that does this that client also uses internally so that we do not need to maintain this code here